### PR TITLE
[Custom Amounts M1] Custom Amount Flow

### LINKED ISSUE #10879: [Custom Amounts M1] Add Custom Amount Flow

### PR DESCRIPTION
**Pull request by toupper**
Sorry for the big PR. This is due to the refactoring to reuse the component existent in Simple Payments, including unit tests.

Closes: #10879 

## Description

With this PR we implement the flow so the merchants can add a custom amount to an order. Here we pass the amount and name to the EditableOrderViewModel, but the actual adding of the amount as a fee to the order will come in a following PR.

We already had a very similar screen for Simple Payments, because of that part of this PR was the extraction of that code to make it reusable.

As you can see, following the example of other view model references in the `EditableOrderViewModel`, we have one reference to the `AddCustomAmountViewModel` instead of recreating it multiple times. The reason is, apart from having to recreate the view model every time we need it, that when rotation the view was recreated sometimes, recreating the view model as well. Because of that, we were losing the state.

## Testing instructions

1. Go to orders
2. Tap on + to create a new order
3. Tap on + Add Custom amount
4. See that the flow is correct. After you save a custom amount, you can see in the Xcode console:

`"Adding custom amount of 12 with name Test"`

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/1864060/19624327-e4bd-4f68-9d95-41684fbdeb12

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountView.swift [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                    .navigationBarItems(leading: Button(action: {
+                        dismiss()
+                    }) {
+                        Text(Localization.navigationCancelButtonTitle)
+                    })
+                }
+                .frame(width: geometry.size.width)
+            }
+        }
+
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountView.swift]:
```suggestion
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+         onCustomAmountEntered: @escaping CustomAmountEntered) {
+        self.formattableAmountTextFieldViewModel = FormattableAmountTextFieldViewModel(locale: locale, storeCurrencySettings: storeCurrencySettings)
+        self.onCustomAmountEntered = onCustomAmountEntered
+        listenToAmountChanges()
+    }
+
+    /// Variable that holds the name of the custom amount.
+    ///
+    @Published var name = ""
+    @Published var shouldDisableDoneButton: Bool = true
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift]: I'm not sure if we rely on this later, so: Should we make these `@Published` properties private and initialize them with the viewmodel? At the moment we seem to read-only access the `name` from the View, so perhaps we only need to expose that and disallow them from being settable.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        onCustomAmountEntered(formattableAmountTextFieldViewModel.amount, customAmountName)
+        reset()
+    }
+
+    func reset() {
+        name = ""
+        shouldDisableDoneButton = true
+
+        formattableAmountTextFieldViewModel.reset()
+    }
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift:40]: Should we separate dismissal from reset? So it's clearer from the EditableViewModel when is triggered:
 ```suggestion
    func dismiss() {
      reset()
    }

    private func reset() {
        name = ""
        shouldDisableDoneButton = true
        formattableAmountTextFieldViewModel.reset()
    }
```

So when we call:
```diff
    func onDismissAddCustomAmountView() {
-        addCustomAmountViewModel.reset()
+         addCustomAmountViewModel.dismiss()
    }
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+private extension AddCustomAmountViewModel {
+    func listenToAmountChanges() {
+        formattableAmountTextFieldViewModel.$amount.map { _ in
+            !self.formattableAmountTextFieldViewModel.amountIsValid
+        }.assign(to: &$shouldDisableDoneButton)
+    }
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift:48]: We seem to do too many things here for the method name, perhaps we could split the method into `binding/listenToAmountChanges` and `validation`?

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/FormattableAmountTextFieldViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import Combine
+import UIKit
+import WooFoundation
+
+final class FormattableAmountTextFieldViewModel: ObservableObject {
+    /// Helper to format price field input.
+    ///
+    let priceFieldFormatter: PriceFieldFormatter
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/FormattableAmountTextFieldViewModel.swift]:
```suggestion
    private let priceFieldFormatter: PriceFieldFormatter
```

File: WooCommerce/Classes/ViewRelated/Orders/Simple Payments/Amount/SimplePaymentsAmountViewModel.swift [hunk 6]
```diff
@@ -150,13 +119,23 @@ final class SimplePaymentsAmountViewModel: ObservableObject {
...
         NewOrderInitialStatusResolver(siteID: siteID, stores: stores).resolve { [weak self] baseStatus in
             self?.initialOrderStatus = baseStatus
         }
     }
+
+    func listenToAmountChanges() {
+        formattableAmountTextFieldViewModel.$amount.map { _ in
+            !self.formattableAmountTextFieldViewModel.amountIsValid
+        }.assign(to: &$shouldDisableDoneButton)
+    }
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Simple Payments/Amount/SimplePaymentsAmountViewModel.swift:138]: Same as above, perhaps we could split the method into `binding/listenToChanges` and `validation`?

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/FormattableAmountTextField.swift [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        .fixedSize(horizontal: false, vertical: true)
+    }
+}
+
+private extension FormattableAmountTextField {
+    enum Layout {
+        static func amountFontSize(scale: CGFloat) -> CGFloat {
+            56 * scale
+        }
+    }
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/FormattableAmountTextField.swift:41]: We can move to the `Layout` enum both the `scale` variable and the `56` multiplier.

**Review by iamgabrielma — APPROVED** [review]

**Comment by iamgabrielma** [comment]
LGTM! As a side note I see a bunch of `'Error: string input is not a number:' ` logs in the console when opening the Custom Amount editor, this seems to come from pre-existing components.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountView.swift (hunk 1 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountView.swift]: Done in 66c31ef3e0ceb82c1c9f8888b5b4ce2b8f30dc6b

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift (hunk 2 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift]: `name` is written on the view, in the text field. `shouldDisableDoneButton` is also accessed from the outside but only to read, I added private(set) in a0eeaea9e533f0dd7b8676408f8ef5c94263683d

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift (hunk 3 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift:40]: I think dismiss and reset are already split, they have their own functions and aren't coupled to each other. Calling `addCustomAmountViewModel.dismiss()` in `onDismissAddCustomAmountView` looks confusing to me, the view is already dismissed; and we don't want to dismiss it again.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift (hunk 4 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift:48]: Thanks for the suggestion! In my opinion, the simplicity of this code is a good trade-off compared to having to add a new method that just assigns the value:

```
func listenToAmountChanges() {
        formattableAmountTextFieldViewModel.$amount.sink(receiveValue: { [weak self] _ in
            self?.updateShouldDisableDoneButton()
        })
        .store(in: &subscriptions)
    }

    func updateShouldDisableDoneButton() {
        shouldDisableDoneButton = formattableAmountTextFieldViewModel.amountIsValid
    }
```

I think the original is simpler and once one is familiar with the reactive syntax of Combine also quite readable.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/FormattableAmountTextFieldViewModel.swift (hunk 5 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/FormattableAmountTextFieldViewModel.swift]: Good catch, done in 742151aca9cebe918449e347da3232c4140da452

**Pull request merged**